`problem:` assuming you work in a hospital and the doctor gives you a patient data to identify if a patient in the given data have heart disease this projects gives you the insight on how to solve that problem as a Machine Learning Practitional
### we are going to be building machine learning model for predicting heart disease
### A Binary Classification Project
### tools we are gonna use for this projects
* Scikit-Learn
* pandas
* numpy
### The model we are going to use would be a neural network which is `MLPClassifier` from the sklearn module

Download the dataset and notebook `Heart Disease Prediction` from this link https://github.com/Akinleyejoshua/j-ml-projects to follow along with this tutorial 

In [2]:
# we are going to first import the libraries for this projects

# pandas is a data manipulation library which can be used to create, manipulate and view DataFrame from datasets
import pandas as pd

# MLPClassifier is a neural network model which would be used to fit in the datasets
from sklearn.neural_network import MLPClassifier

# for converting our data into a train and test datasets, train_test_split module must be imported from the sklearn library
from sklearn.model_selection import train_test_split

# import pandas for numeric computing
import numpy as np

In [5]:
# Import the Heart Disease Datasets and feed it to Pandas DataFrame
heart_disease_df = pd.read_csv("heart-disease.csv")

# List the datasets features and target
heart_disease_df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

### The feature sets are the sets which are independent sets which would be used for prediction
### The target set depends on the feature sets
### The feature sets which in this project would be identified as `X`
### The target sets would be identified as `y`

the feature sets include the `age`, `sex`, `cp` ...
the target sets include `0` and `1`
Since this is a Classification Project the target sets includes the
* `0` means Negative i.e Patient does not have Heart Disease
* `1` means Positive i.e Patient does have Heart Diseases

In [3]:
# describe the dataset using describe function
heart_disease_df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

## as we can see that the count is 303 which means the number of samples is `303`

In [7]:
# store the feature sets into the variable X
# drop function removes an element from a DataFrame
# We do so by removing the target column(axis = 1)
X = heart_disease_df.drop("target", axis=1)
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [8]:
# store the target set into variable y
# we do so by selecting the target column from the dataframe
y = heart_disease_df["target"]
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [9]:
np.unique(heart_disease_df["target"])
# 0 : Negative
# 1: Positive

array([0, 1], dtype=int64)

#### Spliting our datasets into train and test is very important 
* our train datasets takes about 80% while test sets about 20%

#### A train set is used to fit the data into the model
* It also used to allow the model to find pattern in a dataset

#### A test set is used to test our accuracy of our datasets, the better the accuracy the better the learned pattern
* A test set is used to test if the model can generalise on new data which it asn't seen
* The better accuracy the better the model can generalise on new data
* A test set can also serve as a validator to the model

In [10]:
# so we are gonna create a train and test split for X and y
# setting test size to 20% === 0.2
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

### Fitting the `X_train` and `y_train` to the `MLPClassifer` model
* this would make the model learn from the trained split data 
* Pattern are learned from the model

In [13]:
# instantiate the model
model = MLPClassifier()

# fit the data into the model
model.fit(X_train, y_train)

# score the model to get it accuracy
accuracy = model.score(X_test, y_test)
print("Accuracy: ", accuracy)

Accuracy:  0.8524590163934426


C:\ProgramData\Miniconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:614: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


we have an accuracy of about 85% which is also equalls 0.85, which means our model learned very well from the fitted data 
that also means the data can generalize from new data and predict well if a new patient has a heart disease

### Now we have our trained data we can now predict on our test data to check how well the model is doing
1. We would get the `X_test` and make predictions on the data
2. We would compare the `X_test` predictions with `y_test` to check if it actually predicted it accuratelly

In [14]:
# make predictions on X_test
y_pred = model.predict(X_test)
print(y_pred)

[0 1 0 0 1 0 0 1 1 0 1 0 0 1 1 1 0 0 0 1 0 0 0 1 0 1 0 0 1 1 1 0 0 1 1 1 0
 1 0 1 1 0 0 0 1 0 1 0 1 1 1 1 0 1 1 0 0 1 0 0 0]


In [15]:
# compare predictions
boolean_pred = y_pred == y_test
boolean_pred

217     True
177    False
181     True
292     True
113     True
       ...  
172     True
68      True
249     True
194     True
289     True
Name: target, Length: 61, dtype: bool

### As we can see from the predictions that some where predicted `True` and `False`
now let get the amount of predicted `false` and `true` using the `value_counts()` function

In [16]:
boolean_pred.value_counts()

True     52
False     9
Name: target, dtype: int64

wow!!! our model perfomed well, we have more predictions to be accurate(`True`) than `False`
* `52` predictions where `True`
* `9` predictions where `False`

## Now let make predictions on new data the model have'nt seen on the feature sets given from the datasets below:
* `age` `sex` `cp` `trestbps` `chol`  `fbs` `restecg` `thalach` `exang` `oldpeak` `slope` `ca` `thal`
#### assuming the above data shows the patient health status that we want to make prediction on wether the patient have heart disease or not
the code: `name = model.predict([[age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal]])`
### Note: anywhere you see `1` it means `True` and `0` means `False` e.g `sex for 1 means Male which is True`

In [19]:
# now to make it easily understood lets define a variable(the name of the patient) that stores the patient health data
dave = model.predict([[15, 1, 3, 120, 225, 1, 0, 155, 0, 1.5, 0, 0, 3]])
print("Prediction", dave[0])

Prediction 1


Now let create a function that accepts the user data and displays if he or she has heart disease or his positive or negative

Since our prediction from the patient dave is `1 which means True` then `Dave` test result is `Positive` he has heart disease

In [20]:
classes = ["Heart Diease: Negative", "Heart Disease: Positive"]
def get_pred_label(name, pred):
    pred = model.predict([pred])
    return f"{name} : {classes[pred[0]]}"

In [21]:
dave = get_pred_label("Dave", [15, 1, 3, 120, 225, 1, 0, 155, 0, 1.5, 0, 0, 3])
mike = get_pred_label("Mike", [56, 1, 2, 100, 10, 1, 1, 122, 10, 4, 1, 1, 1])
print(dave)
print(mike)

Dave : Heart Disease: Positive
Mike : Heart Diease: Negative


the label result show that dave has heart disease while mike does'nt have heart disease
## Now let store our result in a DataFrame

In [30]:
patients = ["Dave", "Mike"]
results = ["Positive", "Negative"]
patient_data = pd.DataFrame()
patient_data["Patients"] = pd.Series(patients)
patient_data["result"] = pd.Series(results)
patient_data

Patients    result
0     Dave  Positive
1     Mike  Negative